nixos unstable build error: option not set correctly: PREEMPT

I am trying to build musnix (from master) on nixos unstable with kernel.optimize = true (works fine without this)

In my configuration.nix:
```
  musnix = {
    enable = true;
    kernel.optimize = true;
    kernel.packages = pkgs.linuxPackages_latest_rt;
  };
```

The error I get when building the kernel is `error: option not set correctly: PREEMPT (wanted 'y', got 'n')`:
```
GOT: #
GOT: # configuration written to .config
GOT: #
GOT: make[1]: Leaving directory '/build/linux-4.19.48/build'
GOT: make: Leaving directory '/build/linux-4.19.48'
error: option not set correctly: PREEMPT (wanted 'y', got 'n')
builder for '/nix/store/hdig9zrv89q3sjp67xrb84dc55dshis6-linux-config-4.19.48.drv' failed with exit code 255
cannot build derivation '/nix/store/4pz9mahpbkndjcl5kmym9aypv6a0nydj-linux-4.19.48.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/abp3jrk9cy51mzsja0z4vrpf4g5kpd5q-nixos-system-nixos-19.09pre182256.168d1031af4.drv': 1 dependencies couldn't be built
error: build of '/nix/store/abp3jrk9cy51mzsja0z4vrpf4g5kpd5q-nixos-system-nixos-19.09pre182256.168d1031af4.drv' failed
```

Looking further at the output I see it entering scripts/kconfig/conf twice, the first time setting PREEMPT, the second time setting PREEMPT_VOLUNTARY

```
make: Entering directory '/build/linux-4.19.48'
make[1]: Entering directory '/build/linux-4.19.48/build'
  HOSTCC  scripts/basic/fixdep
  GEN     ./Makefile
  HOSTCC  scripts/kconfig/conf.o
  YACC    scripts/kconfig/zconf.tab.c
  LEX     scripts/kconfig/zconf.lex.c
  HOSTCC  scripts/kconfig/zconf.tab.o
  HOSTLD  scripts/kconfig/conf
*** Default configuration is based on 'x86_64_defconfig'
#
# configuration written to .config
#
make[1]: Leaving directory '/build/linux-4.19.48/build'
make: Leaving directory '/build/linux-4.19.48'
generating kernel configuration...
GOT: make: Entering directory '/build/linux-4.19.48'
GOT: make[1]: Entering directory '/build/linux-4.19.48/build'
GOT:   GEN     ./Makefile
GOT: scripts/kconfig/conf  --oldaskconfig Kconfig
GOT: *
GOT: * Linux/x86_64 4.19.48 Kernel Configuration
GOT: *
GOT: *
GOT: * Compiler: gcc (GCC) 7.4.0
GOT: *
GOT: *
GOT: * General setup
GOT: *


...


GOT: Preemption Model
GOT:   1. No Forced Preemption (Server) (PREEMPT_NONE)
GOT: > 2. Voluntary Kernel Preemption (Desktop) (PREEMPT_VOLUNTARY)
GOT:   3. Preemptible Kernel (Low-Latency Desktop) (PREEMPT)
CHOICE: 1-3?, ANSWER: 3


...


GOT: #
GOT: # configuration written to .config
GOT: #
GOT: make[1]: Leaving directory '/build/linux-4.19.48/build'
GOT: make: Leaving directory '/build/linux-4.19.48'
GOT: make: Entering directory '/build/linux-4.19.48'
GOT: make[1]: Entering directory '/build/linux-4.19.48/build'
GOT:   GEN     ./Makefile
GOT: scripts/kconfig/conf  --oldaskconfig Kconfig
GOT: *
GOT: * Linux/x86_64 4.19.48 Kernel Configuration
GOT: *
GOT: *


...


GOT: Preemption Model
GOT:   1. No Forced Preemption (Server) (PREEMPT_NONE)
GOT:   2. Voluntary Kernel Preemption (Desktop) (PREEMPT_VOLUNTARY)
GOT: > 3. Preemptible Kernel (Low-Latency Desktop) (PREEMPT)
CHOICE: 1-3?, ANSWER: 2
```

Let me know if I can provide any other info. Thanks

@lheckemann Thanks!
I tried that and got:
```
 H16TOFW firmware/edgeport/down2.fw
  IHEX    firmware/edgeport/down3.bin
  IHEX2FW firmware/whiteheat_loader.fw
  IHEX2FW firmware/whiteheat.fw
  IHEX2FW firmware/keyspan_pda/keyspan_pda.fw
  IHEX2FW firmware/keyspan_pda/xircom_pgs.fw
  IHEX    firmware/cpia2/stv0672_vp4.bin
  IHEX    firmware/yam/1200.bin
  IHEX    firmware/yam/9600.bin
make[1]: *** [Makefile:150: sub-make] Error 2
make: *** [Makefile:24: __sub-make] Error 2
builder for '/nix/store/xriwl1wrrjr1r0zspkxp2k5ks07m7amh-linux-4.9.35-rt25.drv' failed with exit code 2
cannot build derivation '/nix/store/1p77qifpvx4fjk8djhlaf95a9d3hmmjk-acpi-call-4.9.35-rt25.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/sx16195l450klicqwxkwpmm18597zq6d-firmware.drv': 1 dependencies couldn't be built
building '/nix/store/xmk4k8b5jbgchghqjhlbkrqls7dqy3ja-login.pam.drv'...
cannot build derivation '/nix/store/fhfpbv5ffm4hp624p6mjssdrsjqwndj4-tp_smapi-0.43-4.9.35-rt25.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/rpk06i7dzm1wvcnz1hvzgjazzjvcs6n9-etc-nixos.conf.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/09ggbyh86f07nqhxx9w0q1068rzy7b15-kernel-modules.drv': 3 dependencies couldn't be built
cannot build derivation '/nix/store/5cxirnrgm4yh7bciaavk4x9cl80ki2ns-linux-4.9.35-rt25-modules.drv': 3 dependencies couldn't be built
cannot build derivation '/nix/store/vmzzai4x3pvmmr2hcj1wvfzb94snca0b-linux-4.9.35-rt25-modules-shrunk.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/fvxqj9y2majwg4l9zc5cn35sn0r7whqf-stage-1-init.sh.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/mkjsq8bcnhcqn1fww2w4niz5n2afy96j-initrd-linux-4.9.35-rt25.drv': 1 dependencies couldn't be built
building '/nix/store/388db663xk9j4ym56dp8xlh122w666x6-nixos.conf.drv'...
building '/nix/store/9bxxwgs6cz961l707mmd14n6649sjmxx-passwd.pam.drv'...
building '/nix/store/anr6072726xc73zzvpgl2x0brmgk8bi9-physlock.pam.drv'...
building '/nix/store/m2jkx7rx24455dkpxd7cdmx02a5y1176-polkit-1.pam.drv'...
building '/nix/store/0pshws3fxrnl95id7xzysam05nx6bdaw-rtirq.conf.drv'...
building '/nix/store/yzndgj45g3b6b7ax4afbd8jjrrmjxzag-rtirq.drv'...
```

Any ideas?

@lheckemann Thanks!
Could it be this one?:
```
  VOFFSET arch/x86/boot/compressed/../voffset.h
  CC      arch/x86/boot/string.o
  CC      arch/x86/boot/tty.o
  CC      arch/x86/boot/video.o
  CC      arch/x86/boot/compressed/string.o
  CC      arch/x86/boot/video-mode.o
  CC      arch/x86/boot/version.o
  CC      arch/x86/boot/video-vga.o
  CC      arch/x86/boot/compressed/cmdline.o
  CC      arch/x86/boot/video-vesa.o
  CC      arch/x86/boot/video-bios.o
  CC      arch/x86/boot/compressed/error.o
  OBJCOPY arch/x86/boot/compressed/vmlinux.bin
  HOSTCC  arch/x86/boot/tools/build
  RELOCS  arch/x86/boot/compressed/vmlinux.relocs
Unsupported relocation type: R_X86_64_PLT32 (4)
make[4]: *** [../arch/x86/boot/compressed/Makefile:118: arch/x86/boot/compressed/vmlinux.relocs] Error 1
make[4]: *** Waiting for unfinished jobs....
make[3]: *** [../arch/x86/boot/Makefile:111: arch/x86/boot/compressed/vmlinux] Error 2
make[3]: *** Waiting for unfinished jobs....
make[2]: *** [arch/x86/Makefile:245: bzImage] Error 2
make[2]: *** Waiting for unfinished jobs....
  CC      arch/x86/crypto/aes-x86_64.mod.o
  CC      arch/x86/crypto/aesni-intel.mod.o
  CC      arch/x86/crypto/blowfish-x86_64.mod.o
  CC      arch/x86/crypto/camellia-aesni-avx-x86_64.mod.o
  CC      arch/x86/crypto/camellia-aesni-avx2.mod.o
  CC      arch/x86/crypto/camellia-x86_64.mod.o
  CC      arch/x86/crypto/cast5-avx-x86_64.mod.o
  CC      arch/x86/crypto/cast6-avx-x86_64.mod.o
  CC      arch/x86/crypto/chacha20-x86_64.mod.o
  CC      arch/x86/crypto/crc32-pclmul.mod.o
```

> When I tried to package 5.4 I got the above error again.

Sorry, version 5.4 of what?

> The 5.0 kernel builds and runs for me.

Are you building it declaratively or procedurally? If declarative,
could I see the config you're using to do it?

My [latest config](https://github.com/JeffreyBenjaminBrown/nixos-experiments/blob/audio/audio.nix) builds. It doesn't use realtime. If I add (uncomment) the line `kernel.realtime = true;` it get an "Unsupported relocation type" error, like the one [you mentioned above](https://github.com/musnix/musnix/issues/100#issuecomment-523938192).

```
  CC      arch/x86/boot/regs.o
  CC      arch/x86/boot/string.o
  OBJCOPY arch/x86/boot/compressed/vmlinux.bin
  RELOCS  arch/x86/boot/compressed/vmlinux.relocs
  CC      arch/x86/boot/tty.o
Unsupported relocation type: R_X86_64_PLT32 (4)
make[4]: *** [../arch/x86/boot/compressed/Makefile:118: arch/x86/boot/compressed/vmlinux.relocs]
 Error 1
make[4]: *** Waiting for unfinished jobs....
  CC      arch/x86/boot/video.o
make[3]: *** [../arch/x86/boot/Makefile:111: arch/x86/boot/compressed/vmlinux] Error 2
make[3]: *** Waiting for unfinished jobs....
make[2]: *** [arch/x86/Makefile:245: bzImage] Error 2
make[2]: *** Waiting for unfinished jobs....
  CC      arch/x86/crypto/aes-x86_64.mod.o
  CC      arch/x86/crypto/aesni-intel.mod.o
  CC      arch/x86/crypto/blowfish-x86_64.mod.o
  CC      arch/x86/crypto/camellia-aesni-avx-x86_64.mod.o
  ...
  ... # Here I'm skipping an enormous number of similar-looking lines
  ...
  IHEX    firmware/cpia2/stv0672_vp4.bin
  IHEX    firmware/yam/1200.bin
  IHEX    firmware/yam/9600.bin
make[1]: *** [Makefile:150: sub-make] Error 2
make: *** [Makefile:24: __sub-make] Error 2
builder for '/nix/store/kv3r5mglmil6m6cigqaxjm5cqb5kw0yb-linux-4.9.35-rt25.drv' failed with exit code 2
cannot build derivation '/nix/store/rqwlz520cbrrin6drk82vsh1srh9nkyd-exfat-nofuse-2019-09-06-4.9.35-rt25.drv': 1 dependenc
ies couldn't be built
cannot build derivation '/nix/store/gbrin503x8rn1knh660y72rwskij2gvk-firmware.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/g3c1na4ajj6qlh6cym0f9bp1vy4lipdr-etc-nixos.conf.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/h5pwn6gzhwb16ckqzk20mqsp57q863ll-kernel-modules.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/l698kf4z6b4c4lrvav7y8r72kl7bahv1-linux-4.9.35-rt25-modules.drv': 2 dependencies couldn
't be built
cannot build derivation '/nix/store/wakadccgc19mfmprc4gnyc6qxc3670kv-etc.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/k15zndlrqbbkb7271qray530idhbb47j-linux-4.9.35-rt25-modules-shrunk.drv': 2 dependencies
 couldn't be built
cannot build derivation '/nix/store/y143adn33mimw18bpdcbwlwl8x029dp2-stage-1-init.sh.drv': 1 dependencies couldn't be buil
t
cannot build derivation '/nix/store/hxcji3xbzch182lkpqqv3vl5ycl6vaxq-initrd-linux-4.9.35-rt25.drv': 1 dependencies couldn'
t be built
cannot build derivation '/nix/store/9qwj5rbjz79mx230b4djvyxxgn5i7ssw-nixos-system-jbb-dell-19.09.1889.692a8cabbcc.drv': 5
dependencies couldn't be built
error: build of '/nix/store/9qwj5rbjz79mx230b4djvyxxgn5i7ssw-nixos-system-jbb-dell-19.09.1889.692a8cabbcc.drv' failed
```

--- a/modules/kernel.nix
+++ b/modules/kernel.nix
@@ -1,47 +1,30 @@
 { config, lib, pkgs, ... }:
 
 with lib;
+with lib.kernel;
 
 let
 
   cfg = config.musnix;
 
-  kernelConfigLatencyTOP = ''
-    LATENCYTOP y
-    SCHEDSTATS y
-  '';
+  standardConfig = version: with (lib.kernel.whenHelpers version);
+    optionalAttrs cfg.kernel.latencytop {
+      LATENCYTOP = yes;
+      SCHEDSTATS = yes;
+    } //
+    optionalAttrs cfg.kernel.optimize {
+      PREEMPT = yes;
+      # DEADLINE was renamed to MT_DEADLINE and enabled by default.
+      IOSCHED_DEADLINE = whenOlder "5" yes;
+      DEFAULT_DEADLINE = whenOlder "5" yes;
+      DEFAULT_IOSCHED = whenOlder "5" (freeform "deadline");
+    };
 
-  kernelConfigOptimizeDeadline = ''
-    IOSCHED_DEADLINE y
-    DEFAULT_DEADLINE y
-    DEFAULT_IOSCHED deadline
-  '';
-
-  kernelConfigOptimize = ''
-    HPET_TIMER y
-    TREE_RCU_TRACE? n
-    RT_GROUP_SCHED? n
-  '';
-
-  kernelConfigRealtime = ''
-    PREEMPT y
-    PREEMPT_RT_FULL y
-    RT_GROUP_SCHED? n
-  '';
-
-  musnixRealtimeKernelExtraConfig =
-    kernelConfigRealtime
-    + optionalString cfg.kernel.optimize kernelConfigOptimize
-    + optionalString cfg.kernel.latencytop kernelConfigLatencyTOP;
-
-  musnixStandardKernelExtraConfig =
-    if cfg.kernel.optimize
-      then "PREEMPT y\n"
-           + kernelConfigOptimize
-           + optionalString cfg.kernel.latencytop kernelConfigLatencyTOP
-      else if cfg.kernel.latencytop
-        then kernelConfigLatencyTOP
-        else "";
+  realtimeConfig = version:
+    (standardConfig version) // {
+      PREEMPT = yes;
+      PREEMPT_RT_FULL = yes;
+    };
 
 in {
   options.musnix = {
@@ -66,8 +49,7 @@ in {
         NOTE: Enabling this option will rebuild your kernel.
 
         If enabled, this option will configure the kernel to be
-        preemptible, to use the deadline I/O scheduler, and to use
-        the High Precision Event Timer (HPET).
+        preemptible and use the deadline I/O scheduler.
       '';
     };
     kernel.realtime = mkOption {
@@ -116,21 +98,21 @@ in {
         kernelPatches = [ kernelPatches.bridge_stp_helper
                           realtimePatches.realtimePatch_3_18
                         ];
-        extraConfig   = musnixRealtimeKernelExtraConfig + optionalString cfg.kernel.optimize kernelConfigOptimizeDeadline;
+        structuredExtraConfig = realtimeConfig "3.18";
       };
 
       linux_4_1_rt = callPackage ../pkgs/os-specific/linux/kernel/linux-4.1-rt.nix {
         kernelPatches = [ kernelPatches.bridge_stp_helper
                           realtimePatches.realtimePatch_4_1
                         ];
-        extraConfig   = musnixRealtimeKernelExtraConfig + optionalString cfg.kernel.optimize kernelConfigOptimizeDeadline;
+        structuredExtraConfig = realtimeConfig "4.1";
       };
 
       linux_4_4_rt = callPackage ../pkgs/os-specific/linux/kernel/linux-4.4-rt.nix {
         kernelPatches = [ kernelPatches.bridge_stp_helper
                           realtimePatches.realtimePatch_4_4
                         ];
-        extraConfig   = musnixRealtimeKernelExtraConfig + optionalString cfg.kernel.optimize kernelConfigOptimizeDeadline;
+        structuredExtraConfig = realtimeConfig "4.4";
       };
 
       linux_4_9_rt = callPackage ../pkgs/os-specific/linux/kernel/linux-4.9-rt.nix {
@@ -138,7 +120,7 @@ in {
                           kernelPatches.modinst_arg_list_too_long
                           realtimePatches.realtimePatch_4_9
                         ];
-        extraConfig   = musnixRealtimeKernelExtraConfig + optionalString cfg.kernel.optimize kernelConfigOptimizeDeadline;
+        structuredExtraConfig = realtimeConfig "4.9";
       };
 
       linux_4_11_rt = callPackage ../pkgs/os-specific/linux/kernel/linux-4.11-rt.nix {
@@ -146,7 +128,7 @@ in {
                           kernelPatches.modinst_arg_list_too_long
                           realtimePatches.realtimePatch_4_11
                         ];
-        extraConfig   = musnixRealtimeKernelExtraConfig + optionalString cfg.kernel.optimize kernelConfigOptimizeDeadline;
+        structuredExtraConfig = realtimeConfig "4.11";
       };
 
       linux_4_13_rt = callPackage ../pkgs/os-specific/linux/kernel/linux-4.13-rt.nix {
@@ -154,7 +136,7 @@ in {
                           kernelPatches.modinst_arg_list_too_long
                           realtimePatches.realtimePatch_4_13
                         ];
-        extraConfig   = musnixRealtimeKernelExtraConfig + optionalString cfg.kernel.optimize kernelConfigOptimizeDeadline;
+        structuredExtraConfig = realtimeConfig "4.13";
       };
 
       linux_4_14_rt = callPackage ../pkgs/os-specific/linux/kernel/linux-4.14-rt.nix {
@@ -162,7 +144,7 @@ in {
                           kernelPatches.modinst_arg_list_too_long
                           realtimePatches.realtimePatch_4_14
                         ];
-        extraConfig   = musnixRealtimeKernelExtraConfig + optionalString cfg.kernel.optimize kernelConfigOptimizeDeadline;
+        structuredExtraConfig = realtimeConfig "4.14";
       };
 
       linux_4_16_rt = callPackage ../pkgs/os-specific/linux/kernel/linux-4.16-rt.nix {
@@ -170,34 +152,34 @@ in {
                           kernelPatches.modinst_arg_list_too_long
                           realtimePatches.realtimePatch_4_16
                         ];
-        extraConfig   = musnixRealtimeKernelExtraConfig + optionalString cfg.kernel.optimize kernelConfigOptimizeDeadline;
+        structuredExtraConfig = realtimeConfig "4.16";
       };
       linux_4_18_rt = callPackage ../pkgs/os-specific/linux/kernel/linux-4.18-rt.nix {
         kernelPatches = [ kernelPatches.bridge_stp_helper
                           kernelPatches.modinst_arg_list_too_long
                           realtimePatches.realtimePatch_4_18
                         ];
-        extraConfig   = musnixRealtimeKernelExtraConfig + optionalString cfg.kernel.optimize kernelConfigOptimizeDeadline;
+        structuredExtraConfig = realtimeConfig "4.18";
       };
       linux_4_19_rt = callPackage ../pkgs/os-specific/linux/kernel/linux-4.19-rt.nix {
         kernelPatches = [ kernelPatches.bridge_stp_helper
                           kernelPatches.modinst_arg_list_too_long
                           realtimePatches.realtimePatch_4_19
                         ];
-        extraConfig   = musnixRealtimeKernelExtraConfig + optionalString cfg.kernel.optimize kernelConfigOptimizeDeadline;
+        structuredExtraConfig = realtimeConfig "4.19";
       };
       linux_5_0_rt = callPackage ../pkgs/os-specific/linux/kernel/linux-5.0-rt.nix {
         kernelPatches = [ kernelPatches.bridge_stp_helper
                           kernelPatches.modinst_arg_list_too_long
                           realtimePatches.realtimePatch_5_0
                         ];
-        extraConfig   = musnixRealtimeKernelExtraConfig;
+        structuredExtraConfig = realtimeConfig "5.0";
       };
 
 
 
       linux_opt = linux.override {
-        extraConfig = musnixStandardKernelExtraConfig;
+        structuredExtraConfig = standardConfig linux.version;
       };
 
       linuxPackages_3_18_rt = recurseIntoAttrs (linuxPackagesFor linux_3_18_rt);

